Services/Mail/classes/AutoResponder/AutoResponderServiceImpl.php

```diff
@@ -16,43 +16,61 @@
 
-use ilObjUser;
-use DateTimeImmutable;
 use DateInterval;
 use ilFormatMail;
 use ilMailOptions;
 use ILIAS\Data\Clock\ClockInterface;
-use ILIAS\Data\Factory as DataFactory;
+use DateTimeZone;
+use DateTimeImmutable;
 
-class AutoResponderServiceImpl implements AutoResponderService
+final class AutoResponderServiceImpl implements AutoResponderService
 {
-    protected bool $auto_responder_status;
-    /** @var ilMailOptions[] $auto_responder_data */
-    protected array $auto_responder_data;
-    protected int $global_idle_time_interval;
     /** @var callable  */
-    protected $loginByUsrIdCallable;
-    protected AutoResponderRepository $auto_responder_repository;
-    protected ClockInterface $clock;
+    private $loginByUsrIdCallable;
+    private bool $auto_responder_status;
+    private AutoResponderRepository $auto_responder_repository;
+    private ClockInterface $clock;
+    private int $global_idle_time_interval;
+    /** @var callable */
+    private $mail_action;
+    
+    /** @var ilMailOptions[] $auto_responder_data */
+    protected array $auto_responder_data = [];
 
     public function __construct(
         callable $loginByUsrIdCallable,
         int $global_idle_time_interval,
-        bool $auto_responder_status,
-        array $auto_responder_data,
+        bool $initial_auto_responder_status,
         AutoResponderRepository $auto_responder_repository,
-        ClockInterface $clock
+        ClockInterface $clock,
+        ?callable $mail_action = null
     ) {
+        $this->loginByUsrIdCallable = $loginByUsrIdCallable;
         $this->global_idle_time_interval = $global_idle_time_interval;
-        $this->loginByUsrIdCallable = $loginByUsrIdCallable ?? static function (int $usrId) : string {
-            return ilObjUser::_lookupLogin($usrId);
-        };
-        $this->auto_responder_status = $auto_responder_status;
-        $this->auto_responder_data = $auto_responder_data;
+        $this->auto_responder_status = $initial_auto_responder_status;
         $this->auto_responder_repository = $auto_responder_repository;
-        $this->clock = $clock ?? (new DataFactory())->clock()->utc();
+        $this->clock = $clock;
+        $this->mail_action = $mail_action;
+    }
+
+    private function normalizeDateTimezone(DateTimeImmutable $date_time) : DateTimeImmutable
+    {
+        return $date_time->setTimezone(new DateTimeZone('UTC'));
+    }
+
+    private function shouldSendAutoResponder(AutoResponder $auto_responder) : bool
+    {
+        // Normalize timezones
+        $last_send_time_with_added_interval = $this
+            ->normalizeDateTimezone($auto_responder->getSentTime())
+            ->add(new DateInterval('P' . $this->global_idle_time_interval . 'D'));
+
+        $now = $this->normalizeDateTimezone($this->clock->now());
+
+        // Don't compare the objects because of microseconds
+        return $last_send_time_with_added_interval->format('Y-m-d H:i:s') <= $now->format('Y-m-d H:i:s');
```

Comment:
Wenn wir in Tagen denken für das Intervall, sollten wir dann nicht lieber `H:i:s` aus dem Vergleich rausnehmen?

Reply:
Das ist eine gute Frage.

Szenario:
- Du sendest mir um 09:00 Uhr MESZ 10.08.2022 eine E-Mail
- Ich habe eingestellt, dass der AR nur alle 7 Tage an eine Person verschickt werden soll
- Nun sendest Du um 08:00 Uhr MESZ am 17.08.2022 eine E-Mail
- Die Frage der Fragen ist: Sollst Du einen AR erhalten, ohne nicht.